### PR TITLE
Created DiffieHellman.java and MonoAlphabetic.java 

### PR DESCRIPTION
**Pull request by riti2601**
Description

This PR implements the Diffie-Hellman Key Exchange and MonoAlphabetic Cipher algorithms. 

Key Changes:

1. Diffie-Hellman Key Exchange:
   - Added `DiffieHellman.java` under the `ciphers` directory.
   - The implementation allows two users (Person A and Person B) to securely generate a shared secret key over an insecure channel using large prime numbers and modular arithmetic.
   - The code prompts the users to input their secret numbers and performs the necessary calculations to exchange and compute the shared secret keys.

2. MonoAlphabetic Cipher:
   - Added `MonoAlphabetic.java` under the `ciphers` directory.
   - The implementation allows the user to either **encrypt** or **decrypt** data using a predefined substitution cipher, where each letter of the alphabet is replaced with a corresponding letter.
   - The user is prompted to enter the data, and the cipher is applied to return either the encrypted or decrypted result.

Both algorithms are utility classes, with private constructors to prevent instantiation and static methods to handle their respective functionalities.

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Review by vil02 — CHANGES_REQUESTED** [review]
Please remove the `main` methods and add proper junit tests. Prefer `ParameterizedTest`s.

**Comment by riti2601** [comment]
I am currently unable to build my code due to the outdated pom.xml file. The dependencies required for the JUnit tests are missing, leading to compilation errors. To proceed with testing and building my changes effectively, I recommend updating the pom.xml file to include the necessary JUnit dependencies.

Please let me know if there's a preferred way to address this issue or if I should modify the code to remove the dependencies temporarily.

Thank you!

**Comment by vil02** [comment]
Did you try to place the tests in the [`.../src/test/java/com/thealgorithms`](https://github.com/TheAlgorithms/Java/tree/842ff5294f5e97dfd1e4c09f9d9e01a89ac16e32/src/test/java/com/thealgorithms) directory? cf. #5501.

**Comment by riti2601** [comment]
No, I'll try it. 
Thank you

**Comment by riti2601** [comment]
Hi @vil02,
I've addressed all the requested changes and updated the pull request accordingly. Could you please review the latest changes at your convenience and let me know if anything else is needed?
Thank you!

File: src/test/java/com/thealgorithms/ciphers/DiffieHellmanTest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    // Provide test data for public key calculation
+    private static Stream<Arguments> providePublicKeyData() {
+        return Stream.of(Arguments.of(BigInteger.valueOf(5), BigInteger.valueOf(6), BigInteger.valueOf(23), BigInteger.valueOf(8)), Arguments.of(BigInteger.valueOf(2), BigInteger.valueOf(5), BigInteger.valueOf(13), BigInteger.valueOf(6)));
+    }
+
+    // Provide test data for shared secret calculation
+    private static Stream<Arguments> provideSharedSecretData() {
+        return Stream.of(Arguments.of(BigInteger.valueOf(5), BigInteger.valueOf(6), BigInteger.valueOf(23), BigInteger.valueOf(8), BigInteger.valueOf(13)), Arguments.of(BigInteger.valueOf(2), BigInteger.valueOf(5), BigInteger.valueOf(13), BigInteger.valueOf(6), BigInteger.valueOf(2)));
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/ciphers/DiffieHellmanTest.java:37]: Similarly here.

File: src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package com.thealgorithms.ciphers;
+
+public final class MonoAlphabetic {
+
+    // Private constructor to prevent instantiation of utility class
```
Comment (vil02) [inline src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:5]:
```suggestion
```

File: src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    private MonoAlphabetic() {
+        throw new UnsupportedOperationException("Utility class");
+    }
+
+    // Encryption method
+    public static String encrypt(String data, String key) {
+        StringBuilder sb = new StringBuilder();
+        data = data.toUpperCase();
```
Comment (vil02) [inline src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:13]: I would suggest to throw an exception if the input has lower case letters.

File: src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        for (char c : data.toCharArray()) {
+            if (c >= 'A' && c <= 'Z') {
+                int idx = c - 'A'; // Index in alphabet
```
Comment (vil02) [inline src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:17]: Consider adding two methods: `charToPos` and `posToChar`.

File: src/main/java/com/thealgorithms/ciphers/DiffieHellman.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        // Returns g^x mod p
+        return base.modPow(secret, prime);
+    }
+
+    // Method to calculate the shared secret key (otherPublic^secret mod p)
+    public BigInteger calculateSharedSecret(BigInteger otherPublicValue) {
+        if (otherPublicValue == null || otherPublicValue.signum() <= 0) {
+            throw new IllegalArgumentException("Other public value must be non-null and positive.");
+        }
+        // Returns b^x mod p or a^y mod p
```
Comment (vil02) [inline src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:33]: ?

File: src/main/java/com/thealgorithms/ciphers/DiffieHellman.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            throw new IllegalArgumentException("Base, secret, and prime must be non-null and positive values.");
+        }
+        this.base = base;
+        this.secret = secret;
+        this.prime = prime;
+    }
+
+    // Method to calculate public value (g^x mod p)
+    public BigInteger calculatePublicValue() {
+        // Returns g^x mod p
```
Comment (vil02) [inline src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:24]: What is `g`, `x` and `p`?

**Review by vil02 — DISMISSED** [review]

File: src/main/java/com/thealgorithms/ciphers/DiffieHellman.java (hunk 6 above)
Comment (riti2601) [reply src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:24]: In the Diffie-Hellman key exchange:
1. g represents the base (also called the generator), which is a publicly agreed-upon number used by both parties to perform the key exchange.
2. p is a large prime number, another public value that ensures the security of the key exchange.
3. x is the private secret key chosen by one of the parties. This value is kept secret and used to compute the public key, which is shared with the other party.
Together, these values are used to compute a shared secret without directly transmitting it, ensuring secure communication.

File: src/main/java/com/thealgorithms/ciphers/DiffieHellman.java (hunk 5 above)
Comment (riti2601) [reply src/main/java/com/thealgorithms/ciphers/DiffieHellman.java:33]: These values are used to compute a shared secret without directly transmitting it, ensuring secure communication.

File: src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java (hunk 3 above)
Comment (riti2601) [reply src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:15]: Added the exception accordingly.

File: src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java (hunk 4 above)
Comment (riti2601) [reply src/main/java/com/thealgorithms/ciphers/MonoAlphabetic.java:19]: Added the methods charToPos and posToChar.

**Review by riti2601 — COMMENTED** [review]
Hi @vil02,
I have made the necessary changes according to your suggestions.
Thank you for your valuable feedback!

**Comment by siriak** [comment]
@vil02 please review

**Comment by riti2601** [comment]
Hi @vil02, kindly review the changed code.

**Comment by riti2601** [comment]
Hello @vil02
I am participating in Hacktoberfest'25 and request that you kindly check the PR for further merging. 
Thank you

**Review by siriak — APPROVED** [review]

**Pull request merged**